### PR TITLE
Clear module slots in DCT when preparing for an OTA update

### PR DESCRIPTION
**Pull request by sergeuz**
### Problem

When system resets are disabled via `System.disableReset()`, it is possible to flash the device OTA multiple times without it resetting itself to apply the first received update. Module slots that are stored in the DCT and trigger the bootloader to copy the module's data from the OTA section to its destination section in the internal flash are not cleared in this case, even if the received modules occupy the same area in the OTA section, which may result in a corrupted firmware update.

### Solution

Clear all non-factory module slots in the DCT at the beginning of an OTA update.

### Steps to Test

1. Flash the example application to a Gen 2 or Gen 3 device. Let the device connect to the cloud.
2. Flash the main system part module that is already running on the device OTA. The device will not reset after the update and stay connected to the cloud.
3. Flash some firmware module that has unmet dependencies to the device OTA. One way to get such a module is to bump the `BOOTLOADER_DEPENDENCY` version of the main system part module that is running on the device.
4. Reset the device via the reset button and verify that the bootloader doesn't try to apply any updates. Running this test without the fix in place would result in the bootloader copying the module with unmet dependencies to its destination section in the internal flash.

### Example App

```c
#include "application.h"

SYSTEM_MODE(SEMI_AUTOMATIC)
SYSTEM_THREAD(ENABLED)

void setup() {
    System.disableReset();
    Particle.connect();
}

void loop() {
}
```

### References

- [ch84957]

### CODE REVIEW

**Review by avtolstoy — APPROVED** [review]

**Comment by avtolstoy** [comment]
Looks good, we have a path forward for an automated test for this as well but dependent on hal_system_info changes.

File: platform/MCU/STM32F2xx/SPARK_Firmware_Driver/src/flash_mal.c [hunk 1]
```diff
@@ -873,16 +873,43 @@ uint32_t FLASH_PagesMask(uint32_t imageSize, uint32_t pageSize)
...
+    const bool ok = FLASH_EraseMemory(FLASH_SERIAL, FLASH_Address, imageSize);
 #else
-    FLASH_EraseMemory(FLASH_INTERNAL, FLASH_Address, imageSize);
+    const bool ok = FLASH_EraseMemory(FLASH_INTERNAL, FLASH_Address, imageSize);
 #endif
+    if (!ok) {
+        return FLASH_ACCESS_RESULT_ERROR;
+    }
+
+    return 0;
```
Comment (technobly) [inline platform/MCU/STM32F2xx/SPARK_Firmware_Driver/src/flash_mal.c:912]: Suggestion: `return FLASH_ACCESS_RESULT_OK;`

File: platform/MCU/nRF52840/src/flash_mal.c [hunk 2]
```diff
@@ -952,16 +952,43 @@ void FLASH_Restore(uint32_t FLASH_Address)
...
+    const bool ok = FLASH_EraseMemory(FLASH_SERIAL, FLASH_Address, imageSize);
 #else
-    FLASH_EraseMemory(FLASH_INTERNAL, FLASH_Address, imageSize);
+    const bool ok = FLASH_EraseMemory(FLASH_INTERNAL, FLASH_Address, imageSize);
 #endif
+    if (!ok) {
+        return FLASH_ACCESS_RESULT_ERROR;
+    }
+
+    return 0;
```
Comment (technobly) [inline platform/MCU/nRF52840/src/flash_mal.c:991]: Suggestion: `return FLASH_ACCESS_RESULT_OK;`

**Review by technobly — APPROVED** [review]
Looks good, Sergey!  Clearing this on every chunk received seems necessary since an async reset can happen at any time.  Nice error handling ;-)

**Comment by sergeuz** [comment]
@technobly The slots are cleared just once at the beginning of the update. Clearing them on every received chunk is not necessary as they're populated with the module details only after the module binary has been fully received and validated in `HAL_FLASH_End()`.

**Comment by technobly** [comment]
@sergeuz Ah, I see that now... that's even better, thanks :D

File: system/src/system_cloud_internal.cpp [hunk 3]
```diff
@@ -1079,12 +1079,12 @@ void Spark_Protocol_Init(void)
...
 
         // Enable device-initiated describe messages
-        spark_protocol_set_connection_property(sp, protocol::Connection::DEVICE_INITIATED_DESCRIBE, 0, nullptr, nullptr);
+        spark_protocol_set_connection_property(sp, protocol::Connection::ENABLE_DEVICE_INITIATED_DESCRIBE, 0, nullptr, nullptr);
 
 #if HAL_PLATFORM_COMPRESSED_OTA
         // Enable compressed/combined OTA updates
         if (bootloader_get_version() >= COMPRESSED_OTA_MIN_BOOTLOADER_VERSION) {
-            spark_protocol_set_connection_property(sp, protocol::Connection::COMPRESSED_OTA, 0, nullptr, nullptr);
+            spark_protocol_set_connection_property(sp, protocol::Connection::COMPRESSED_OTA, 1, nullptr, nullptr);
```
Comment (sergeuz) [inline system/src/system_cloud_internal.cpp:1087]: This is a safe change as compressed updates are only supported on Gen 3 where the comms library and the system are in the same system part.

**Comment by sergeuz** [comment]
I added an integration test for this fix: `ota/multiple_ota_no_reset`. The test runner 0.3.13 is required in order to run it.

**Comment by avtolstoy** [comment]
Tested for now manually on Tracker and Electron :heavy_check_mark:

**Pull request merged**